### ISSUE TITLE
Multiple lodash

### ISSUE THREAD

**Issue by xarg** [algolia/instantsearch.js#984]
If you already have lodash in your package:

```
{
  "name": "algolia-test",
  "version": "1.0.0",
  "description": "",
  "main": "index.js",
  "scripts": {
    "test": "echo \"Error: no test specified\" && exit 1"
  },
  "author": "",
  "license": "ISC",
  "dependencies": {
    "lodash": "^4.11.1"
  }
}
```

then when you `npm install instantsearch.js` it will download lodash 3 times again:

```
$ ls node_modules/**/node_modules/
node_modules/algoliasearch-helper/node_modules/:
lodash

node_modules/algoliasearch/node_modules/:
lodash

node_modules/instantsearch.js/node_modules/:
lodash
```

Ok.. this is how npm does things when there are version conflicts, but.. then when you're building with webpack it imports all lodash again for every package (almost 2mb redundant stuff in total):

```
...
 [891] ./~/algoliasearch-helper/~/lodash/math/sum.js 1.35 kB {0} [built]
  [892] ./~/algoliasearch-helper/~/lodash/internal/arraySum.js 491 bytes {0} [built]
  [893] ./~/algoliasearch-helper/~/lodash/internal/baseSum.js 569 bytes {0} [built]
  [894] ./~/algoliasearch-helper/~/lodash/internal/toIterable.js 527 bytes {0} [built]
  [895] ./~/algoliasearch-helper/~/lodash/object/values.js 691 bytes {0} [built]
  [896] ./~/algoliasearch-helper/~/lodash/internal/baseValues.js 593 bytes {0} [built]
  [897] ./~/algoliasearch-helper/~/lodash/collection/includes.js 2.05 kB {0} [built]
  [898] ./~/algoliasearch-helper/~/lodash/collection/sortByOrder.js 2.06 kB {0} [built]
  [899] ./~/algoliasearch-helper/~/lodash/internal/baseSortByOrder.js 1.1 kB {0} [built]
  [900] ./~/algoliasearch-helper/~/lodash/internal/baseSortBy.js 543 bytes {0} [built]
  [901] ./~/algoliasearch-helper/~/lodash/internal/compareMultiple.js 1.62 kB {0} [built]
  [902] ./~/algoliasearch-helper/~/lodash/internal/baseCompareAscending.js 1.06 kB {0} [built]
  [903] ./~/algoliasearch-helper/~/lodash/function/partial.js 1.29 kB {0} [built]
  [904] ./~/algoliasearch-helper/~/lodash/internal/createPartial.js 619 bytes {0} [built]
  [905] ./~/algoliasearch-helper/~/lodash/function/partialRight.js 1.26 kB {0} [built]
  [906] ./~/algoliasearch-helper/src/functions/formatSort.js 587 bytes {0} [built]
  [907] ./~/algoliasearch-helper/src/SearchResults/generate-hierarchical-tree.js 5.09 kB {0} [built]
  [908] ./~/algoliasearch-helper/~/lodash/string/trim.js 1.35 kB {0} [built]
  [909] ./~/algoliasearch-helper/~/lodash/internal/charsLeftIndex.js 543 bytes {0} [built]
  [910] ./~/algoliasearch-helper/~/lodash/internal/charsRightIndex.js 516 bytes {0} [built]
  [911] ./~/algoliasearch-helper/~/lodash/internal/trimmedLeftIndex.js 501 bytes {0} [built]
  [912] ./~/algoliasearch-helper/~/lodash/internal/isSpace.js 639 bytes {0} [built]
  [913] ./~/algoliasearch-helper/~/lodash/internal/trimmedRightIndex.js 474 bytes {0} [built]
  [914] ./~/algoliasearch-helper/~/lodash/object/pick.js 1.51 kB {0} [built]
  [915] ./~/algoliasearch-helper/src/requestBuilder.js 9.41 kB {0} [built]
  [916] ./~/util/util.js 15.6 kB {0} [built]
  [917] ./~/util/support/isBufferBrowser.js 203 bytes {0} [built]
  [918] ./~/algoliasearch-helper/src/url.js 5.83 kB {0} [built]
  [919] ./~/algoliasearch-helper/src/SearchParameters/shortener.js 2.09 kB {0} [built]
  [920] ./~/algoliasearch-helper/~/lodash/object/invert.js 1.58 kB {0} [built]
  [921] ./~/qs/lib/index.js 200 bytes {0} [built]
  [922] ./~/qs/lib/stringify.js 3.91 kB {0} [built]
  [923] ./~/qs/lib/utils.js 4.32 kB {0} [built]
  [924] ./~/qs/lib/parse.js 5.36 kB {0} [built]
  [925] ./~/algoliasearch-helper/~/lodash/object/mapKeys.js 781 bytes {0} [built]
  [926] ./~/algoliasearch-helper/~/lodash/internal/createObjectMapper.js 736 bytes {0} [built]
  [927] ./~/algoliasearch-helper/~/lodash/object/mapValues.js 1.61 kB {0} [built]
  [928] ./~/algoliasearch-helper/src/version.js 41 bytes {0} [built]
  [929] ./~/instantsearch.js/~/lodash/collection/forEach.js 1.38 kB {0} [built]
  [930] ./~/instantsearch.js/~/lodash/internal/arrayEach.js 534 bytes {0} [built]
  [931] ./~/instantsearch.js/~/lodash/internal/baseEach.js 489 bytes {0} [built]
  [932] ./~/instantsearch.js/~/lodash/internal/baseForOwn.js 475 bytes {0} [built]
  [933] ./~/instantsearch.js/~/lodash/internal/baseFor.js 608 bytes {0} [built]
  [934] ./~/instantsearch.js/~/lodash/internal/createBaseFor.js 709 bytes {0} [built]
  [935] ./~/instantsearch.js/~/lodash/internal/toObject.js 308 bytes {0} [built]
  [936] ./~/instantsearch.js/~/lodash/lang/isObject.js 732 bytes {0} [built]
...
```

Here's a repo with a POC: https://github.com/xarg/instantsearch.js-984

```
npm install
webpack --progress --display-modules
```

**Comment by vvo** [comment]
Ok so I have been digging a bit more this issue and yes we are duplicating lodash.

To sump up:
- you are using lodash v4
- we are using lodash v3 (one duplicate)
- [client](https://github.com/algolia/algoliasearch-client-js) and [helper](https://github.com/algolia/algoliasearch-helper-js) are using lodash v3 BUT [asking](https://github.com/algolia/algoliasearch-helper-js/blob/4357cf95d7efb0c1d017a0c897093289bf31c266/package.json#L48) [for](https://github.com/algolia/algoliasearch-client-js/blob/151a8ea8d079a5fc404d41dd53ae15d42ff020f4/package.json#L9) lodash-compat for old IES compatibility
- instantsearch.js is not asking for lodash-compat

What we end up with is:
- when you require('algoliasearch') you get lodash-compat v3
- when you require('instantsearch.js') you get lodash v3
- when you require('lodash'), since you are asking for v4, you get v4

This is a subject we will tackle soon.

Since the lodash code is very similar in compat/not-compat I would keep using webpack because when gzipped, your code size should not be that much impacted by v3/v3-compat duplicates.

Then in your own code if you want to avoid duplicates, stick to lodash v3.

Let me know what you think. And yes, this is an issue we need to fix.

**Comment by xarg** [comment]
@vvo I thought that a naive solution to this is to create a 'master' package that contains all the instasearch.js deps vendorized inside the same package, this way you'll have only one lodash for everything. I'm not so experienced with NPM though to know if it's possible to do.

As a quick fix, yes we could use an older lodash, but v4 is nicer and the future so I'll stick with that in my project.

Concerning the size, gzip is going to make a lot of duplication go away, but webpack still has to parse those files every time when restarting the dev server/building etc.. which makes it slower. I hate slow.

**Comment by bobylito** [comment]
we could use peer dependencies on lodash. The only problem would be the incompatibilities in the libraries that could occur... moving from v3 to v4 of a library in not usually without consequences and could break.

**Comment by Haroenv** [comment]
We're using lodash v4 since a while now, this should not really be an issue anymore. Feel free to reopen if you still have that issue with the latest version

**Comment by corentin-gautier** [comment]
Hi !

Still having the issue : 
![image](https://user-images.githubusercontent.com/4660731/46605617-6e8a6b00-cafa-11e8-86ed-9290810b2236.png)

Is there a way to solve this ? (I don't import lodash myself, meaning it's not referenced explicitly in my package.json)

Thanks !

**Comment by Haroenv** [comment]
@francoischalifour, can you take a look? @corentin-gautier, can you list your package.json and how you import files? I think another module might force a too tight version of lodash (or maybe it's InstantSearch, but we need to investigate it). 

When you are using Yarn, you can hotfix this by forcing a single resolved field like this: https://yarnpkg.com/en/docs/selective-version-resolutions#search

Can you also say what `yarn why lodash` gives on your app?

**Comment by corentin-gautier** [comment]
``` json
  "dependencies": {
    "@agm/core": "^1.0.0-beta.3",
    "@angular/animations": "^6.1.4",
    "@angular/common": "^6.1.4",
    "@angular/compiler": "^6.1.4",
    "@angular/core": "^6.1.4",
    "@angular/forms": "^6.1.4",
    "@angular/http": "^6.1.4",
    "@angular/platform-browser": "^6.1.4",
    "@angular/platform-browser-dynamic": "^6.1.4",
    "@angular/platform-server": "^6.1.7",
    "@angular/router": "^6.1.4",
    "@ngu/carousel": "1.4.8",
    "@nguniversal/common": "^6.1.0",
    "@nguniversal/express-engine": "^6.1.0",
    "@nguniversal/module-map-ngfactory-loader": "^6.1.0",
    "@ngx-translate/core": "^10.0.2",
    "@ngx-translate/http-loader": "^3.0.1",
    "@ngx-universal/translate-loader": "^5.0.0",
    "angular-instantsearch": "^2.0.4",
    "classlist.js": "^1.1.20150312",
    "core-js": "^2.5.4",
    "hammerjs": "^2.0.8",
    "preboot": "^6.0.0-beta.6",
    "rxjs": "^6.0.0",
    "rxjs-compat": "^6.2.2",
    "zone.js": "~0.8.26"
  },
```

I don't use yarn on my project, I use webpack (it's a angular universal project)

**Comment by corentin-gautier** [comment]
@Haroenv in my package-lock.json the dependency of lodash for instantsearch.js is set to "version": "4.17.5"

**Comment by Haroenv** [comment]
@tkrugg, can you change the lodash version in Angular InstantSearch to be the same as InstantSearch.js?

**Comment by corentin-gautier** [comment]
@Haroenv @tkrugg Could you provide a release date for this ? I need to go on production really soon, would be nice to have a smaller bundle size :/

**Comment by francoischalifour** [comment]
Hey @corentin-gautier, you might have heard that [GitHub is currently experiencing issues](https://status.github.com/messages). Today is probably not a good day to release a new version. We'll do it as soon as the situation is more stable (hopefully tomorrow).

**Comment by francoischalifour** [comment]
Fix released in version [2.10.3](https://github.com/algolia/instantsearch.js/blob/develop/CHANGELOG.md#2103-2018-10-29) (2018-10-29).

Closing this issue.

**Comment by Haroenv** [comment]
Please open a new issue with complete reproducible example if you also see this happening with a latest version of InstantSearch
